resilience4j-micronaut/src/main/java/io/github/resilience4j/mirconaut/bulkhead/BulkheadInterceptor.java

```diff
@@ -0,0 +1,128 @@
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A {@link MethodInterceptor} that intercepts all method calls which are annotated with a {@link io.github.resilience4j.mirconaut.annotation.Bulkhead}
+ * annotation.
+ **/
+@Singleton
+@Requires(beans = BulkheadRegistry.class)
+public class BulkheadInterceptor extends BaseInterceptor implements MethodInterceptor<Object,Object>{
+
+    private final BulkheadRegistry bulkheadRegistry;
+    private final BeanContext beanContext;
+
+    /**
+     *
+     * @param beanContext The bean context to allow for DI of class annotated with {@link javax.inject.Inject}.
+     * @param bulkheadRegistry bulkhead registry used to retrieve {@link Bulkhead} by name
+     */
+    public BulkheadInterceptor(BeanContext beanContext,
+                               BulkheadRegistry bulkheadRegistry) {
+        this.bulkheadRegistry = bulkheadRegistry;
+        this.beanContext = beanContext;
+    }
+
+    @Override
+    public int getOrder() {
+        return ResilienceInterceptPhase.BULKHEAD.getPosition();
+    }
+
+    /**
+     * Finds a fallback method for the given context.
+     *
+     * @param context The context
+     * @return The fallback method if it is present
+     */
+    @Override
+    public Optional<? extends MethodExecutionHandle<?, Object>> findFallbackMethod(MethodInvocationContext<Object, Object> context) {
+        ExecutableMethod executableMethod = context.getExecutableMethod();
+        final String fallbackMethod = executableMethod.stringValue(io.github.resilience4j.mirconaut.annotation.Bulkhead.class, "fallbackMethod").orElse("");
+        Class<?> declaringType = context.getDeclaringType();
+        return beanContext.findExecutionHandle(declaringType, fallbackMethod, context.getArgumentTypes());
+    }
+
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+
+        Optional<AnnotationValue<io.github.resilience4j.mirconaut.annotation.Bulkhead>> opt = context.findAnnotation(io.github.resilience4j.mirconaut.annotation.Bulkhead.class);
+        if (!opt.isPresent()) {
+            return context.proceed();
+        }
+        final io.github.resilience4j.mirconaut.annotation.Bulkhead.Type type = opt.get().enumValue("type", io.github.resilience4j.mirconaut.annotation.Bulkhead.Type.class).orElse(io.github.resilience4j.mirconaut.annotation.Bulkhead.Type.SEMAPHORE);
+        if (type != io.github.resilience4j.mirconaut.annotation.Bulkhead.Type.SEMAPHORE) {
+            return context.proceed();
+        }
+
+        final String name = opt.get().stringValue("name").orElse("default");
+        Bulkhead bulkhead = this.bulkheadRegistry.bulkhead(name);
+        ReturnType<Object> rt = context.getReturnType();
+        Class<Object> returnType = rt.getType();
+        if (CompletionStage.class.isAssignableFrom(returnType)) {
+            Object result = context.proceed();
```

Comment:
Yes, correct. 
Please see https://github.com/resilience4j/resilience4j/blob/master/resilience4j-spring/src/main/java/io/github/resilience4j/bulkhead/configure/BulkheadAspect.java#L221-L230

Reply:
The result it is used in an executeCompletionStage on line 104. I write it to an object first to do a null check. so would it just execute the proceeded within the executeCompletionStage? if its null then it will error out right? I guess it shouldn't matter because it will still throw it back to the fallbackCompletable. 

```
return this.fallbackCompletable(bulkhead.executeCompletionStage(() -> ((CompletableFuture<?>) context.proceed())),context);
```